File: backend/app/api/v1/payments.py
"""收款管理 API。"""
from typing import Annotated, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import PermissionChecker
from app.database import get_db
from app.models.user import User
from app.schemas.payment import (
    PaymentClaimCreate,
    PaymentConfirmRequest,
    PaymentCreate,
    PaymentListOut,
    PaymentOut,
    PaymentStatsOut,
    PaymentUpdate,
)
from app.services import payment as payment_service

router = APIRouter(prefix="/payments", tags=["收款管理"])


@router.get("/stats", response_model=PaymentStatsOut, summary="收款统计")
def stats(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["payment:view"]))],
) -> PaymentStatsOut:
    return PaymentStatsOut(**payment_service.payment_stats(db, current_user))


@router.get("", response_model=PaymentListOut, summary="收款列表")
def list_payments(
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["payment:view"]))],
    status: Optional[str] = None,
    due_status: Optional[str] = Query(None, description="not_due/due_soon/due/overdue/settled"),
    contract_id: Optional[int] = None,
    record_type: Optional[str] = Query(None, description="plan/claim"),
    keyword: Optional[str] = None,
    scope: Optional[str] = Query(None, description="mine/all"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
) -> PaymentListOut:
    total, items = payment_service.list_payments(
        db,
        current_user,
        status=status,
        due_status=due_status,
        contract_id=contract_id,
        record_type=record_type,
        keyword=keyword,
        scope_filter=scope,
        page=page,
        page_size=page_size,
    )
    return PaymentListOut(total=total, items=[PaymentOut.model_validate(x) for x in items])


@router.post("", response_model=PaymentOut, summary="登记应收/收款计划")
def create_payment(
    payload: PaymentCreate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["payment:view"]))],
) -> PaymentOut:
    payment = payment_service.create_payment(db, current_user, payload)
    return PaymentOut.model_validate(payment)


@router.post("/claims", response_model=PaymentOut, summary="提交到款认领")
def create_claim(
    payload: PaymentClaimCreate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["payment:view"]))],
) -> PaymentOut:
    payment = payment_service.create_claim(db, current_user, payload)
    return PaymentOut.model_validate(payment)


@router.get("/{payment_id}", response_model=PaymentOut, summary="收款详情")
def get_payment(
    payment_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["payment:view"]))],
) -> PaymentOut:
    payment = payment_service.get_payment(db, current_user, payment_id)
    return PaymentOut.model_validate(payment)


@router.patch("/{payment_id}", response_model=PaymentOut, summary="编辑待收款")
def update_payment(
    payment_id: int,
    payload: PaymentUpdate,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["payment:view"]))],
) -> PaymentOut:
    payment = payment_service.update_payment(db, current_user, payment_id, payload)
    return PaymentOut.model_validate(payment)


@router.post("/{payment_id}/confirm", response_model=PaymentOut, summary="确认到账/核销认领")
def confirm_payment(
    payment_id: int,
    payload: PaymentConfirmRequest,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["payment:view"]))],
) -> PaymentOut:
    payment = payment_service.confirm_payment(db, current_user, payment_id, payload)
    return PaymentOut.model_validate(payment)


@router.post("/{payment_id}/refund", response_model=PaymentOut, summary="退款")
def refund_payment(
    payment_id: int,
    db: Annotated[Session, Depends(get_db)],
    current_user: Annotated[User, Depends(PermissionChecker(["payment:view"]))],
    reason: Optional[str] = None,
) -> PaymentOut:
    payment = payment_service.refund_payment(db, current_user, payment_id, reason)
    return PaymentOut.model_validate(payment)

File: frontend/src/api/payments.ts
/**
 * 收款管理 API
 */
import request from './request'

export type PaymentStatus = 'pending' | 'pending_review' | 'confirmed' | 'refunded'
export type PaymentRecordType = 'plan' | 'claim'

export interface Payment {
  id: number
  payment_no?: string | null
  contract_id: number
  record_type?: PaymentRecordType | string
  title?: string | null
  amount: number | string
  due_date?: string | null
  paid_date?: string | null
  status: PaymentStatus | string
  method?: string | null
  payer_name?: string | null
  account_tail?: string | null
  proof_filename?: string | null
  owner_id?: number | null
  creator_id?: number | null
  department_id?: number | null
  confirmed_by?: number | null
  confirmed_at?: string | null
  remark?: string | null
  created_at: string
  updated_at: string
  contract_no?: string | null
  contract_title?: string | null
  customer_id?: number | null
  customer_name?: string | null
  owner_name?: string | null
  creator_name?: string | null
  confirmed_by_name?: string | null
  due_status?: string | null
}

export interface PaymentStats {
  total: number
  pending: number
  confirmed: number
  refunded: number
  overdue: number
  pending_amount: number | string
  confirmed_amount: number | string
  mine: number
  pending_review?: number
  pending_review_amount?: number | string
  due_soon_amount?: number | string
  month_contract_amount?: number | string
  collection_rate?: number | string
  forecast_gross_margin?: number | string
}

export const PAYMENT_STATUS_LABEL: Record<string, string> = {
  pending: '待收款',
  pending_review: '待复核',
  confirmed: '已核销',
  refunded: '已退款',
}

export const DUE_STATUS_LABEL: Record<string, string> = {
  pending: '待收款',
  not_due: '未到期',
  due_soon: '即将到期',
  due: '已到期',
  overdue: '逾期',
  settled: '已结清',
  refunded: '已退款',
  pending_review: '待复核',
}

export const PAYMENT_METHOD_OPTIONS = [
  { value: 'bank', label: '银行转账' },
  { value: 'alipay', label: '支付宝' },
  { value: 'wechat', label: '微信' },
  { value: 'cash', label: '现金' },
  { value: 'other', label: '其他' },
]

export function fetchPaymentStats() {
  return request.get<PaymentStats>('/payments/stats')
}

export function fetchPayments(params: {
  status?: string
  due_status?: string
  contract_id?: number
  record_type?: string
  keyword?: string
  scope?: string
  page?: number
  page_size?: number
}) {
  return request.get<{ total: number; items: Payment[] }>('/payments', { params })
}

export function fetchPaymentDetail(id: number) {
  return request.get<Payment>(`/payments/${id}`)
}

export function createPayment(data: {
  contract_id: number
  title?: string
  amount: number
  due_date?: string
  method?: string
  remark?: string
}) {
  return request.post<Payment>('/payments', data)
}

export function createPaymentClaim(data: {
  contract_id: number
  amount: number
  paid_date: string
  payer_name: string
  account_tail?: string
  proof_filename: string
  remark?: string
}) {
  return request.post<Payment>('/payments/claims', data)
}

export function updatePayment(id: number, data: Partial<Payment>) {
  return request.patch<Payment>(`/payments/${id}`, data)
}

export function confirmPayment(
  id: number,
  data?: { paid_date?: string; method?: string; remark?: string },
) {
  return request.post<Payment>(`/payments/${id}/confirm`, data || {})
}

export function refundPayment(id: number, reason?: string) {
  return request.post<Payment>(`/payments/${id}/refund`, null, { params: { reason } })
}
